dotenv-linter/src/checks/substitution_key.rs

```diff
@@ -44,7 +44,11 @@ impl Check for SubstitutionKeyChecker<'_> {
             let is_incorrect_substitution = has_start_brace ^ has_end_brace
                 || end_brace_index
                     .map(|i| &initial_key[1..i])
-                    .filter(|key| key.contains(|c: char| !c.is_ascii_alphanumeric() && c != '_'))
+                    .filter(|key| {
+                        key.contains(|c: char| {
+                            !c.is_ascii_alphanumeric() && !":-+=?".contains(c) && c != '_'
```

Comment:
`!":-+=?".contains(c) && c != '_'` 

Looks like it can be merged to `!":-+=?_".contains(c)`

Reply:
another suggestion:
We could define a static `HashSet` to slightly improve the perf.  However it requires either LazyCell (stabilized recently) or lazy_static

E.g:

```rust
// or `LazyLock`
const ALLOWED_CHARS: LazyCell<HashSet<char>> = LazyCell::new(|| HashSet::from([':', '-', '+', '=', '?', '_']));
// ...
&& !ALLOWED_CHARS.contains(c)
```

Reply:
But hold on, please
First of all, we need to figure out our expectations here: https://github.com/dotenv-linter/dotenv-linter/pull/786/files#r1625018876

Reply:
But hold on, please
First of all, we need to figure out our expectations here: https://github.com/dotenv-linter/dotenv-linter/pull/786/files#r1625018876

Reply:
These suggestions don't make sense, because the approach should be changed